hibernate-core/src/main/antlr/org/hibernate/grammars/hql/HqlParser.g4

```diff
@@ -35,38 +35,40 @@ subQuery
 	: querySpec
 	;
 
+rootEntity
+	: entityName identificationVariableDef?
+	;
+
 deleteStatement
-	: DELETE FROM? entityName identificationVariableDef? whereClause?
+	: DELETE FROM? rootEntity whereClause?
 	;
 
 updateStatement
-	: UPDATE FROM? entityName identificationVariableDef? setClause whereClause?
+	: UPDATE rootEntity setClause whereClause?
 	;
 
 setClause
-	: SET assignment+
+	: SET assignment (COMMA assignment)*
 	;
 
 assignment
 	: dotIdentifierSequence EQUAL expression
 	;
 
 insertStatement
-// todo (6.0 : VERSIONED
-	: INSERT insertSpec querySpec
+	: INSERT INTO? rootEntity targetFieldsSpec (querySpec | valuesList)
 	;
 
-insertSpec
-	: intoSpec targetFieldsSpec
+targetFieldsSpec
+	: LEFT_PAREN dotIdentifierSequence (COMMA dotIdentifierSequence)* RIGHT_PAREN
 	;
 
-intoSpec
-	: INTO entityName
+valuesList
+	: VALUES values (COMMA values)*
 	;
 
-targetFieldsSpec
-	:
-	LEFT_PAREN dotIdentifierSequence (COMMA dotIdentifierSequence)* RIGHT_PAREN
+values
+	: LEFT_PAREN expression (COMMA expression)* RIGHT_PAREN
```

Comment:
I know very little about ANTLR. Is it possible to dicate the above two list share the same number of elements? Sorry for my ignorance.

Reply:
No, it's not possible, and you generally don't want to do as little validation as possible in the parser because parsers don't usually give good error messages.

However, you're right that we should have code to detect user error here. However that is a problem that predates my work (it also affects `insert ... select ...`).